homeassistant/components/tuya/climate.py

```diff
@@ -142,10 +142,10 @@ def __init__(
         # If both temperature values for celsius and fahrenheit are present,
         # use whatever the device is set to, with a fallback to celsius.
         prefered_temperature_unit = None
-        if all(
+        if any(
```

Comment:
This change is incorrect; it only affects "preferred" temperature, in case both are present.

The actual temperature data point to use is defined/figured out on line 159-192

Reply:
Ah this makes sense to what it was trying to do.  The problem is we need `prefered_temperature_unit` to be set or else the rest fails if only farhenheit is present as:
https://github.com/home-assistant/core/blob/442c5ccc0634a40ae2fa75eb51218ea68e3331bf/homeassistant/components/tuya/climate.py#L169-L171   requires the preferred unit to be set to something other than None to enter that if block, and https://github.com/home-assistant/core/blob/442c5ccc0634a40ae2fa75eb51218ea68e3331bf/homeassistant/components/tuya/climate.py#L186-L188  again requires it to be set to something other than None to enter that if block.
WIthout one of those being entered _set_temperature has no chance to be set to farhenheit_type.

I think the current PR actually still works in all situations, there are only 4 conditions going in:

- no functions F or C are available - the PR code path operates the same as the original code path here (as any is not satisfied just as all was not)
- Only F is available - The any condition now passes, this results in `prefered_temperature_unit = TEMP_CELSIUS` no matter what.  If TEMP_UNIT_CONVERT and DPCode.C_F are both not avail preferred is never set to `TEMP_FAHRENHEIT` but it would not have been before.  The only different now is prefered_temperature_unit is TEMP_CELSIUS rather than None, but that will only change the rest of the flow if celcius functions are not present (which would seem to be desired).
- Only C is available -  Code flow remains the same as celcius_type is defined so functions are set to C
- Both F & C versions are available - the PR code path operates the same as the original code path here (as both the any and all versions are satisfied)

The better solution may be to actually default to preferring Celsius by default, and only changing if farhenheit is available  and celsius is not or if both are in F is preferred.   

So I believe the solution will be https://github.com/home-assistant/core/blob/442c5ccc0634a40ae2fa75eb51218ea68e3331bf/homeassistant/components/tuya/climate.py#L144
 set to TEMP_CELSIUS rather than none by default.   That will allow the conditions like `or (prefered_temperature_unit == TEMP_CELSIUS and not celsius_type)` to fail properly.

I will go ahead and do a build with that change and see if it also resolves the issue.